webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/data/NodeDetailLevel.scala

```diff
@@ -159,61 +161,63 @@ object NodeDetailLevel {
-    val inventoryDate: NodeInfo => JValue = (inv: NodeInfo) => DateFormaterService.getFormatedDate(inv.inventoryDate)
-    val properties   : NodeInfo => JValue = (inv: NodeInfo) => inv.properties.toApiJson
-    val policyMode   : NodeInfo => JValue = (inv: NodeInfo) => inv.policyMode.map(_.name).getOrElse[String]("default")
-    val timezone     : NodeInfo => JValue = (inv: NodeInfo) => inv.timezone.map( t => ("name" -> t.name) ~ ("offset" -> t.offset) )
+  type INFO = (NodeInfo, Option[DateTime])
+  private val nodeInfoFields: Map[String, INFO => JValue] = {
+
+    val id           : INFO => JValue = (info:INFO) => info._1.id.value
+    val hostname     : INFO => JValue = (info:INFO) => info._1.hostname
+    val state        : INFO => JValue = (info:INFO) => info._1.state.name
+    val description  : INFO => JValue = (info:INFO) => info._1.description
+    val policyServer : INFO => JValue = (info:INFO) => info._1.policyServerId.value
+    val ram          : INFO => JValue = (info:INFO) => info._1.ram.map(MemorySize.sizeMb)
+    val arch         : INFO => JValue = (info:INFO) => info._1.archDescription
+    val runDate      : INFO => JValue = (info:INFO) => info._2.map(d => JString(DateFormaterService.getFormatedDate(d))).getOrElse(JNothing)
+    val inventoryDate: INFO => JValue = (info:INFO) => DateFormaterService.getFormatedDate(info._1.inventoryDate)
+    val properties   : INFO => JValue = (info:INFO) => info._1.properties.toApiJson
+    val policyMode   : INFO => JValue = (info:INFO) => info._1.policyMode.map(_.name).getOrElse[String]("default")
+    val timezone     : INFO => JValue = (info:INFO) => info._1.timezone.map( t => ("name" -> t.name) ~ ("offset" -> t.offset) )
 
     val os = {
-      ( inv : NodeInfo ) =>
-        val osType = inv.osDetails.os match {
+      ( info : INFO ) =>
+        val osType = info._1.osDetails.os match {
           case _:LinuxType   => "Linux"
           case _:WindowsType => "Windows"
           case _:BsdType     => "BSD"
           case AixOS         => "AIX"
           case SolarisOS     => "Solaris"
           case UnknownOSType => "Unknown"
         }
-        ( "type" -> osType ) ~
-        ( "name" -> inv.osDetails.os.name ) ~
-        ( "version"  -> inv.osDetails.version.value ) ~
-        ( "fullName" -> inv.osDetails.fullName ) ~
-        ( "servicePack"   -> inv.osDetails.servicePack ) ~
-        ( "kernelVersion" -> inv.osDetails.kernelVersion.value )
+        ( "type"          -> osType ) ~
+        ( "name"          -> info._1.osDetails.os.name ) ~
+        ( "version"       -> info._1.osDetails.version.value ) ~
+        ( "fullName"      -> info._1.osDetails.fullName ) ~
+        ( "servicePack"   -> info._1.osDetails.servicePack ) ~
+        ( "kernelVersion" -> info._1.osDetails.kernelVersion.value )
     }
 
     val machine = {
-      ( inv : NodeInfo ) =>
-        val (machineType,provider) = inv.machine.map(_.machineType match {
+      ( info : INFO ) =>
```

Comment:
Here we could have unapplied the info to a tuple (inv, optLastRunDate ) but no pb il like the _1 too and would be quite PITA for you to rewrite it !!